gateway/ln-gateway/src/state_machine/pay.rs

```diff
@@ -242,20 +251,89 @@ impl GatewayPayInvoice {
         }
     }
 
+    async fn buy_preimage_via_direct_swap(
+        client: Client,
+        invoice: Bolt11Invoice,
+        contract: OutgoingContractAccount,
+    ) -> Result<Preimage, OutgoingPaymentError> {
+        let swap_params = invoice
+            .try_into()
+            .map_err(|e| OutgoingPaymentError::SwapFailed {
+                contract: contract.clone(),
+                swap_error: format!("Failed to parse invoice into swap parameters: {}", e),
+            })?;
+        let operation_id = client
+            .gateway_handle_direct_swap(swap_params)
+            .await
+            .map_err(|e| OutgoingPaymentError::SwapFailed {
+                contract: contract.clone(),
+                swap_error: format!("Failed to initiate direct swap: {}", e),
+            })?;
+        let mut preimage_sub = client
+            .gateway_subscribe_ln_receive(operation_id)
+            .await
+            .map_err(|e| OutgoingPaymentError::SwapFailed {
+                contract: contract.clone(),
+                swap_error: format!("Failed to subscribe to direct swap: {}", e),
+            })?
+            .into_stream();
+
+        while let Some(state) = preimage_sub.next().await {
```

Comment:
I don't think we want to wait on the output of the receive state machine in the transition function. If we detect that this should be an internal swap, I think we should spawn the state machine, but transition to a new state that explicitly waits on the state of the child sm in the trigger function. We do this in `complete.rs` as well.

We don't want to block in the transition function ideally since we're holding a lock and no other state machines can make progress. The only reason we do `pay` for external lightning payments in the transition function is because `pay` is not idempotent.